### ISSUE TITLE
Setting dtypes to keep from promoting to float64 and Apple MPS error

### ISSUE THREAD

**Issue by jeremyg19** [tom-andersson/deepsensor#171]
Originally this was posted in a [neuralprocesses issue](https://github.com/wesselb/neuralprocesses/issues/28#issuecomment-4623475417) but was decided to move to deepsensor. Here's the issue copied over: 

No worries at all in the delay and thanks for replying. I was toying a bit with this and setting the dtype_lik at the call to nps.loglik would be happening in the deepsensor/model/convnp.py file which is a bit "behind the scenes" and to my understanding not really at a point where user inputs would get passed. 

I tried instantiating my ConvNP model with `dtype_lik=torch.float32` or `dtype=torch.float32` which gives a `unexpected keyword` warning and later fails here:

```python
File "/Users/jagraha/dev/repos/deepsensor/deepsensor/train/train.py", line 168, in __init__
    self.opt = optim.Adam(model.model.parameters(), lr=lr)
```

What caught my eye was on the deepsensor end in the def loss_fn function (line 892 in deepsensor/model/convnp.py) the model_kwargs that get passed into `backend.nps.loglik(...)` come from the return of `convert_task_to_nps_arg` on line 919

`context_data, xt, yt, model_kwargs = convert_task_to_nps_args(task)` only returns back the `Y_t_aux` kwarg 

```
model_kwargs = {}
    if "Y_t_aux" in task.keys():
        model_kwargs["aux_t"] = task["Y_t_aux"]

    return context_data, xt, yt, model_kwargs
```

which when passed through causes the promotion to float64 and MPS error. The good is that updating the `model_kwargs` manually at this point

```python
model_kwargs
{'aux_t': tensor([[[-0.8296, -0.8290, -0.8268,  ..., -0.8190, -0.8159, -0.8044]]],
       device='mps:0')}
model_kwargs.update({'dtype_lik': torch.float32})
model_kwargs 
{'aux_t': tensor([[[-0.8296, -0.8290, -0.8268,  ..., -0.8190, -0.8159, -0.8044]]],
       device='mps:0'), 'dtype_lik': torch.float32}
```

then passing on to `backend.nps.loglik(...)` works. 

So seems perhaps the better fix would be updating `convert_task_to_nps_args` to return the dtype_lik in model_kwargs but I'm not entirely sure how best to do that without ConvNP taking in a `dtype_lik` kwarg. 

TL;DR
updating the model_kwargs in deepsensor/model/convnp.py and leaving /neuralprocesses/model/loglik.py alone fixes the issue

```diff
+ 33 import torch
  919 context_data, xt, yt, model_kwargs = convert_task_to_nps_args(task)
+ 920 model_kwargs.update({'dtype_lik': torch.float32}) 
```
 How to best move that dict update from a kwarg passed on the Model instantiation rather than manually here would likely address the issue.

**Comment by davidwyld** [comment]
Thanks @jeremyg19 for reporting this, again sorry for the slow reply.

This looks like a solution, though I'm reluctant to modify the behaviour for all cases and am reluctant to import torch in this module since it damages the backend flexibility.

Presumably you are calling `deepsensor.train.train_epoch` which is calling `loss_fn`? I'm wondering if we can pass some `kwargs` down from the public bits of the API so this type can be set by the user.

It would help if you were able to provide a minimal example for me to test a solution on. Is that possible?

**Comment by jeremyg19** [comment]
Hi David, no worries about a slow reply as mine is even slower! I agree with your thought about not wanting to implement a torch import in the middle of the module, it was a pretty hacky way for me to keep pushing towards my own objectives. 

Indeed I it comes from the eventual call of `loss_fn` in `deepsensor.train.train_epoch` on line 118 (if using torch)
`task_losses.append(model.loss_fn(task, normalise=True))`

which then goes into `deepsensor.convnp.loss_fn` which is where `model_kwargs` gets returned from this call on 919
`context_data, xt, yt, model_kwargs = convert_task_to_nps_args(task)` which instantiates an empty dict, populates it with `aux_t` and returns. 

I'll have to backburner for the immediate time being to make minimal reproducible example as I'm having some difficulty wrapping my head around how to create it now that I've been making local changes to work on the Mac. Maybe it's best to start in a new environment with a new clone of deepsensor, but to even get to that point I had to make a change way back in `trainer.py` in order to utilize the Mac MPS backend. If just using cpu there's no issue with the promotion to float64 and no error gets triggered. 

Perhaps it would be helpful to separate this into to issues:

1) Adding to trainer to allow for mps backend
2) Allow for a dtype model_kwarg to be set and passed on from model instantiation

I think that if 1) was resolved it might be easier to come up with a small reproducible example just using an example ERA5 dataset on a mac. Sorry I'm not able to make this as straightforward as would be helpful. I'll 

#### Changes to `trainer.py` to set the `mps` backend:
```diff
diff --git a/deepsensor/train/train.py b/deepsensor/train/train.py
index 2d9405e..f48b0b4 100644
--- a/deepsensor/train/train.py
+++ b/deepsensor/train/train.py
@@ -12,33 +12,37 @@ from typing import List
 def set_gpu_default_device() -> None:
     """Set default GPU device for the backend.
 
-    Raises:
-        RuntimeError
-            If no GPU is available.
-        RuntimeError
-            If backend is not supported.
-        NotImplementedError
-            If backend is not supported.
-
-    Returns:
-        None.
+    Supports CUDA and Apple MPS for the PyTorch backend.
     """
     if deepsensor.backend.str == "torch":
-        # Run on GPU if available
         import torch
 
         if torch.cuda.is_available():
-            # Set default GPU device
-            torch.set_default_device("cuda")
-            B.set_global_device("cuda:0")
+            device = "cuda"
+            lab_device = "cuda:0"
+
+        elif (
+            hasattr(torch.backends, "mps")
+            and torch.backends.mps.is_built()
+            and torch.backends.mps.is_available()
+        ):
+            device = "mps"
+            lab_device = "mps"
+
         else:
-            raise RuntimeError("No GPU available: torch.cuda.is_available() == False")
+            raise RuntimeError(
+                "No GPU available for torch backend. "
+                "torch.cuda.is_available() == False and "
+                "torch.backends.mps.is_available() == False."
+            )
+
+        torch.set_default_device(device)
+        B.set_global_device(lab_device)
+
```
